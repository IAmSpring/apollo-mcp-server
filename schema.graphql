scalar Date
scalar DateTime

type Query {
  reservation(confirmationNumber: String!): Reservation
  searchHotels(
    location: String!
    checkIn: Date!
    checkOut: Date!
    guests: Int
    roomType: String
  ): [Hotel!]!
  hotels: [Hotel!]!
  hotel(id: ID!): Hotel
}

type Mutation {
  createBooking(input: CreateBookingInput!): Reservation!
  updateReservation(input: UpdateReservationInput!): Reservation!
  cancelReservation(confirmationNumber: String!): Reservation!
}

input CreateBookingInput {
  guestName: String!
  email: String!
  phone: String!
  checkIn: Date!
  checkOut: Date!
  roomType: String!
  numberOfGuests: Int!
  hotelId: ID!
  specialRequests: String
}

input UpdateReservationInput {
  confirmationNumber: String!
  guestName: String
  email: String
  phone: String
  checkIn: Date
  checkOut: Date
  roomType: String
  numberOfGuests: Int
  specialRequests: String
}

type Reservation {
  id: ID!
  confirmationNumber: String!
  guestName: String!
  email: String!
  phone: String!
  checkIn: Date!
  checkOut: Date!
  roomType: String!
  numberOfGuests: Int!
  status: ReservationStatus!
  totalAmount: Float!
  currency: String!
  hotel: Hotel!
  specialRequests: String
  createdAt: DateTime!
  updatedAt: DateTime!
}

type Hotel {
  id: ID!
  name: String!
  address: String!
  city: String!
  state: String!
  country: String!
  zipCode: String!
  phone: String!
  email: String!
  website: String!
  rating: Float!
  reviewCount: Int!
  amenities: [String!]!
  images: [String!]!
  availableRooms: Int!
  priceRange: PriceRange!
  distance: Float
  coordinates: Coordinates
}

type PriceRange {
  min: Float!
  max: Float!
  currency: String!
}

type Coordinates {
  latitude: Float!
  longitude: Float!
}

enum ReservationStatus {
  CONFIRMED
  PENDING
  CANCELLED
  COMPLETED
  NO_SHOW
} 