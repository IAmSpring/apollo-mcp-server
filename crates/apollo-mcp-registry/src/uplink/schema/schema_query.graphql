query SupergraphSdlQuery(
    $apiKey: String!,
    $graphRef: String!,
    $ifAfterId: ID
) {
    routerConfig(
        ref: $graphRef,
        apiKey: $apiKey,
        ifAfterId: $ifAfterId
    ) {
        __typename
        ... on RouterConfigResult {
            id
            supergraphSDL
            minDelaySeconds
        }
        ... on Unchanged {
            id
            minDelaySeconds
        }
        ... on FetchError {
            code
            message
        }
    }
}
