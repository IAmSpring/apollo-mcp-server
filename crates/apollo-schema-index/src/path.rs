use apollo_compiler::ast::NamedType;
use std::borrow::Cow;
use std::fmt;
use std::fmt::Display;
use std::hash::Hash;

/// A path to a type in a schema, starting from a root type
#[derive(Clone, PartialEq, Eq, Hash)]
pub struct RootPath<'a> {
    pub types: Vec<Cow<'a, NamedType>>,
}

impl<'a> RootPath<'a> {
    /// Create a new root path from the given types
    pub fn new(types: Vec<&'a NamedType>) -> Self {
        Self {
            types: types.into_iter().map(Cow::Borrowed).collect(),
        }
    }

    /// Extend this path by adding a new type to the end
    pub fn extend(&self, next_type: &'a NamedType) -> Self {
        let mut types = self.types.clone();
        types.push(Cow::Borrowed(next_type));
        Self { types }
    }

    /// Determine if this path contains a cycle
    pub fn has_cycle(&self) -> bool {
        if let Some(last_type) = self.types.last() {
            self.types
                .get(0..self.types.len() - 1)
                .map(|slice| slice.contains(last_type))
                .unwrap_or(false)
        } else {
            false
        }
    }

    /// Get the type immediately referencing the leaf type
    pub fn referencing_type(&self) -> Option<&NamedType> {
        if self.types.len() > 1 {
            self.types.get(self.types.len() - 2).map(|t| t.as_ref())
        } else {
            None
        }
    }
}

impl<'a> Display for RootPath<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            self.types
                .iter()
                .map(|t| t.as_str())
                .collect::<Vec<_>>()
                .join(" -> ")
        )
    }
}

/// An item with a score
pub struct Scored<T: Eq + Hash + Display> {
    pub inner: T,
    score: f32,
}

impl<T: Eq + Hash + Display> Scored<T> {
    /// Create a new scored item
    pub fn new(inner: T, score: f32) -> Self {
        Self { inner, score }
    }

    /// Get the score associated with this item
    pub fn score(&self) -> f32 {
        self.score
    }
}

impl<T: Eq + Hash + Display> PartialEq for Scored<T> {
    fn eq(&self, other: &Self) -> bool {
        self.inner == other.inner && self.score() == other.score()
    }
}

impl<T: Eq + Hash + Display> Eq for Scored<T> {}

impl<T: Eq + Hash + Display> PartialOrd for Scored<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl<T: Eq + Hash + Display> Ord for Scored<T> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.score().total_cmp(&other.score())
    }
}

impl<T: Eq + Hash + Display> Hash for Scored<T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.inner.hash(state);
    }
}

impl<T: Eq + Hash + Display> Display for Scored<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({})", self.inner, self.score)
    }
}
