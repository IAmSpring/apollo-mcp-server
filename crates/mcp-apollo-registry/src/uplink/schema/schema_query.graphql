query SupergraphSdlQuery(
    $apiKey: String!,
    $graphRef: String!,
    $ifAfterId: ID
) {
    routerConfig(
        ref: $graphRef,
        apiKey: $apiKey,
        ifAfterId: $ifAfterId
    ) {
        __typename
        ... on RouterConfigResult {
            id
            supergraphSdl
            minDelaySeconds
        }
        ... on Unchanged {
            id
            minDelaySeconds
        }
        ... on FetchError {
            code
            message
        }
    }
}
