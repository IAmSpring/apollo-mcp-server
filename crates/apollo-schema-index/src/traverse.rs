//! Provides an extension trait for traversing GraphQL schemas, using a depth-first traversal
//! starting at the specified root operation types (query, mutation, subscription).

use crate::OperationType;
use crate::path::RootPath;
use apollo_compiler::Schema;
use apollo_compiler::ast::NamedType;
use apollo_compiler::schema::ExtendedType;
use enumset::EnumSet;
use itertools::Itertools;
use std::collections::HashMap;
use std::collections::hash_map::Entry;

/// Extension trait to allow traversing a schema
pub trait SchemaExt {
    /// Traverse the type hierarchy in the schema in depth-first order, starting with the specified
    /// root operation types
    fn traverse(
        &self,
        root_types: EnumSet<OperationType>,
    ) -> Box<dyn Iterator<Item = (&ExtendedType, RootPath)> + '_>;
}

impl SchemaExt for Schema {
    fn traverse(
        &self,
        root_types: EnumSet<OperationType>,
    ) -> Box<dyn Iterator<Item = (&ExtendedType, RootPath)> + '_> {
        let mut stack = vec![];
        let mut references: HashMap<&NamedType, Vec<NamedType>> = HashMap::default();
        for root_type in root_types
            .iter()
            .rev()
            .filter_map(|rt| self.root_operation(rt.into()))
        {
            stack.push((root_type, RootPath::new(vec![root_type])));
        }
        Box::new(std::iter::from_fn(move || {
            while let Some((named_type, current_path)) = stack.pop() {
                if current_path.has_cycle() {
                    continue;
                }
                let references = references.entry(named_type);

                // Only traverse the children of a type the first time we visit it.
                // After that, we still visit unique paths to the type, but not the child paths.
                let traverse_children: bool = matches!(references, Entry::Vacant(_));

                references.or_insert(
                    current_path
                        .referencing_type()
                        .map(|t| vec![t.clone()])
                        .unwrap_or_default(),
                );
                if let Some(extended_type) = self.types.get(named_type) {
                    if !extended_type.is_built_in() {
                        if traverse_children {
                            match extended_type {
                                ExtendedType::Object(obj) => {
                                    stack.extend(
                                        obj.fields
                                            .values()
                                            .map(|field| &field.ty)
                                            .map(|ty| ty.inner_named_type())
                                            .unique()
                                            .map(|next_type| {
                                                (next_type, current_path.extend(next_type))
                                            }),
                                    );
                                    stack.extend(
                                        obj.fields
                                            .values()
                                            .flat_map(|field| &field.arguments)
                                            .map(|arg| arg.ty.inner_named_type())
                                            .unique()
                                            .map(|next_type| {
                                                (next_type, current_path.extend(next_type))
                                            }),
                                    );
                                }
                                ExtendedType::Interface(interface) => {
                                    stack.extend(
                                        interface
                                            .fields
                                            .values()
                                            .map(|field| &field.ty)
                                            .map(|ty| ty.inner_named_type())
                                            .unique()
                                            .map(|next_type| {
                                                (next_type, current_path.extend(next_type))
                                            }),
                                    );
                                    stack.extend(
                                        interface
                                            .fields
                                            .values()
                                            .flat_map(|field| &field.arguments)
                                            .map(|arg| arg.ty.inner_named_type())
                                            .unique()
                                            .map(|next_type| {
                                                (next_type, current_path.extend(next_type))
                                            }),
                                    );
                                }
                                ExtendedType::InputObject(input) => {
                                    stack.extend(
                                        input
                                            .fields
                                            .values()
                                            .map(|field| &field.ty)
                                            .map(|ty| ty.inner_named_type())
                                            .unique()
                                            .map(|next_type| {
                                                (next_type, current_path.extend(next_type))
                                            }),
                                    );
                                }
                                ExtendedType::Union(union) => {
                                    stack.extend(
                                        union.members.iter().map(|member| &member.name).map(
                                            |next_type| (next_type, current_path.extend(next_type)),
                                        ),
                                    );
                                }
                                _ => {}
                            }
                        }
                        return Some((extended_type, current_path));
                    }
                }
            }
            None
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use apollo_compiler::validation::Valid;
    use rstest::{fixture, rstest};

    const TEST_SCHEMA: &str = include_str!("testdata/schema.graphql");

    #[fixture]
    fn schema() -> Valid<Schema> {
        Schema::parse(TEST_SCHEMA, "schema.graphql")
            .expect("Failed to parse test schema")
            .validate()
            .expect("Failed to validate test schema")
    }

    #[rstest]
    fn test_schema_traverse(schema: Valid<Schema>) {
        let mut paths = vec![];
        for (_extended_type, path) in schema
            .traverse(OperationType::Query | OperationType::Mutation | OperationType::Subscription)
        {
            paths.push(path.to_string());
        }
        insta::assert_debug_snapshot!(paths);
    }
}
