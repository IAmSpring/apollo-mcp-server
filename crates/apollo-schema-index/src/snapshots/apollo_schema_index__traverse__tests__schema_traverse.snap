---
source: crates/apollo-schema-index/src/traverse.rs
expression: paths
---
[
    "Query",
    "Query -> NotificationFilter",
    "Query -> NotificationFilter -> DateTime",
    "Query -> NotificationFilter -> NotificationPriority",
    "Query -> PostFilter",
    "Query -> PostFilter -> DateRangeInput",
    "Query -> PostFilter -> DateRangeInput -> DateTime",
    "Query -> PostFilter -> ContentStatus",
    "Query -> SearchResult",
    "Query -> SearchResult -> Tag",
    "Query -> SearchResult -> Tag -> Post",
    "Query -> SearchResult -> Tag -> Post -> PostAnalytics",
    "Query -> SearchResult -> Tag -> Post -> PostAnalytics -> Demographics",
    "Query -> SearchResult -> Tag -> Post -> PostAnalytics -> Demographics -> DeviceStats",
    "Query -> SearchResult -> Tag -> Post -> PostAnalytics -> Demographics -> LocationStats",
    "Query -> SearchResult -> Tag -> Post -> PostAnalytics -> Demographics -> AgeGroup",
    "Query -> SearchResult -> Tag -> Post -> Media",
    "Query -> SearchResult -> Tag -> Post -> Media -> User",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserPreferences",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserPreferences -> PrivacySettings",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserPreferences -> NotificationPreferences",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Notification",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Notification -> JSON",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Notification -> DateTime",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Notification -> NotificationPriority",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Comment",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Comment -> Reaction",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Comment -> Reaction -> DateTime",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> Comment -> DateTime",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserProfile",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserProfile -> Location",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserProfile -> Location -> Coordinates",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserProfile -> SocialLink",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> UserRole",
    "Query -> SearchResult -> Tag -> Post -> Media -> User -> DateTime",
    "Query -> SearchResult -> Tag -> Post -> Media -> MediaMetadata",
    "Query -> SearchResult -> Tag -> Post -> Media -> MediaMetadata -> Dimensions",
    "Query -> SearchResult -> Tag -> Post -> Media -> MediaType",
    "Query -> SearchResult -> Tag -> Post -> Comment",
    "Query -> SearchResult -> Tag -> Post -> JSON",
    "Query -> SearchResult -> Tag -> Post -> User",
    "Query -> SearchResult -> Tag -> Post -> ContentStatus",
    "Query -> SearchResult -> Tag -> Post -> DateTime",
    "Query -> SearchResult -> Comment",
    "Query -> SearchResult -> Post",
    "Query -> SearchResult -> User",
    "Query -> Notification",
    "Query -> Comment",
    "Query -> Post",
    "Query -> User",
    "Query -> Node",
    "Query -> Node -> DateTime",
    "Mutation",
    "Mutation -> UserPreferencesInput",
    "Mutation -> UserPreferencesInput -> PrivacySettingsInput",
    "Mutation -> UserPreferencesInput -> NotificationPreferencesInput",
    "Mutation -> Upload",
    "Mutation -> CreateCommentInput",
    "Mutation -> UpdatePostInput",
    "Mutation -> UpdatePostInput -> ContentStatus",
    "Mutation -> CreatePostInput",
    "Mutation -> CreatePostInput -> Upload",
    "Mutation -> CreatePostInput -> ContentStatus",
    "Mutation -> CreateUserInput",
    "Mutation -> CreateUserInput -> CreateUserProfileInput",
    "Mutation -> CreateUserInput -> CreateUserProfileInput -> CreateLocationInput",
    "Mutation -> CreateUserInput -> CreateUserProfileInput -> CreateLocationInput -> CreateCoordinatesInput",
    "Mutation -> CreateUserInput -> UserRole",
    "Mutation -> UserPreferences",
    "Mutation -> Media",
    "Mutation -> Comment",
    "Mutation -> Post",
    "Mutation -> User",
    "Subscription",
    "Subscription -> Notification",
    "Subscription -> Comment",
    "Subscription -> Post",
]
